src/Compilers/Core/Portable/Compilation/RebuildData.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class RebuildData
+    {
+        private Queue<string> NonSourceFileDocumentNames { get; } = new();
```

Comment:
Is this queue expected to be very small or empty for most compilations? Should we use an ArrayBuilder or something internally and expose a Free() method on this instance?

Reply:
I'll flip to a builder here. The number of entries here roughly correlates with items like number of XAML, Razor, etc ... files in a project. That is the most common use case I'm aware of. Enough that it's worth doing a builder.